### ISSUE TITLE
[Feature] / [Question]

### ISSUE THREAD

**Issue by ufst-ask** [microsoft/playwright#27650]
The browsers has a really good text finder built in (the one activated by "ctrl + f"). I would very much like to be able to use this in Playwright as well. Including selecting a specific occurence of a phrase searched for.

I can't see that it's currently possible. Is it? If not, is it something that could be added? I believe it would be very powerful at times. Both to click on items and assert them.

**My attempts and results**
I've tried:
`await expect(page.getByText('Welcome, John')).toBeVisible();` from [docs](https://playwright.dev/docs/locators#:~:text=You%20can%20locate%20the%20element%20by%20the%20text%20it%20contains%3A)
And:
`await expect(page.locator("text=modal title")).toBeVisible()` suggested by Max Schmitt [here](https://stackoverflow.com/questions/70179441/check-if-element-is-visible-in-playwright)

The latter seems to work best, but when I get more than one occurence it breaks.
I'm trying to save them in a list but it won't allow me to. I get
"Error: locator.all: Execution context was destroyed, most likely because of a navigation"
Eventually I want to click certain element but for now this is my code:
`for(const element of await page.locator("text=Sagsbehandler - godkender").all()) {
    await element.click();
  }`

**Comment by mxschmitt** [comment]
> The latter seems to work best, but when I get more than one occurence it breaks.

To make sure there is more than one occurrence, you can do the following:

`await expect(page.getByText('Welcome, John').first()).toBeVisible();`

> Eventually I want to click certain element but for now this is my code:

In your end-to-end test, do you expect the user to also click on all the occurrences of a specific button e.g.?

I'm curious to hear more about your use-case, because it seems like there is an existing solution available rather than introducing a new feature.

**Comment by ufst-ask** [comment]
Sorry for the late response, I made the mistake of asking this question right before I went on holiday.

The "first()" worked to get proper occurence, so thank you for that :)

However, I still have an issue with finding elements by text.
I experience that it finds the text but still times out:
![image](https://github.com/microsoft/playwright/assets/77072723/46d652ec-ef9f-4cf9-b4ca-7c016badfeb6)
![image](https://github.com/microsoft/playwright/assets/77072723/bcfbf3df-b36a-4542-8f50-a1c8b6d6fc23)
(I'm not a 100% sure I can share any code specific to our project, so I censured the unique string).
Any idea on why that is and how to fix it?

I agree that it seems there is an existing solution but I think I'm misunderstanding something about the locators or such.

My use-case is that we're using Playwright for an old system. It's PSRM by Oracle and I think it's more than 20 years old. And Playwright still seems to be the best tool for us (though you focus on "modern websites").

For now I think our best strategy for recognising elements would be:
1. Automate with codegen. When that fails then:
2. Find by text-search on whole page. When that fails then:
3. Develop customized solution which we place in our "project library" for re-use.

I have found that I can navigate most of PSRM simply by using the browsers built in search function. But I have yet to make it work with Playwright.

I hope I made it clear.

**Comment by mxschmitt** [comment]
> I experience that it finds the text but still times out:

This means in most of the cases that the element with the specific text is not on the page. You can use e.g. our Trace Viewer to find out how the page looked like, when the timeout has happened. Or you can increase the timeout when your text might take some more time to appear. (e.g. the site/backend is slow).

> But I have yet to make it work with Playwright.

These APIs are not exposed to JavaScript, so the browser (or we) can't re-use the logic. I would strongly recommend to read over our [Locators guide](https://playwright.dev/docs/locators) and find there an appropriate locator for your element. See also here about our [best-practises](https://playwright.dev/docs/best-practices) which contain useful information about how to locate an element.

**Comment by mxschmitt** [comment]
Could we step back a little bit and come back to the actual problem which you are trying to solve?

Do you try to assert an element by the text on the page? Do you try to click on an element on the page by text?

Why does this not work for you?

```ts
await expect(page.getByText('Welcome, John').first()).toBeVisible();
await page.getByText('Welcome, John').first().click();
```

The timeout error is pretty clear in my opinion, when the timeout has been reached, there is no element with the specific text / locator on the page. Solution is to either increase the timeout or debug the test using trace viewer / UI Mode.

We would love to understand the problem in more detail first before we explore potential solutions. Thanks!

**Comment by ufst-ask** [comment]
No problem
I don't mind spending time on my problems, I'm just glad you're taking your time to help me :)

I wanted to see how much I could navigate just by searching the entire page by text. So in that specific example where I got a timeout, I could just increase the timeout and use codegen instead. Since codegen figured out that it was an iframe, in an iframe in an iframe... 

However, I do still encounter other places where timeout and codegen is not the solution. Here I probably could find a customized solution as a developer. But what I want to achieve is a backup to codegen where non-developer colleagues can find there own solution on such occations. Hence the find_in_page by phrase.

An example of such code is:
`await page.frame({
    name: 'zoneMapFrame_1'
  }).getByRole('link', { name: 'View case' }).click();`
For this I get the error: "Object is possibly 'null'.ts(2531)"
And yes I could probably debug this using trace viewer / UI Mode. But I'm hoping to achieve something even easier to use. Don't get me wrong, those options are a great, just not for a non-developer.

**Comment by mxschmitt** [comment]
> await page.frame({ name: 'zoneMapFrame_1' }).getByRole('link', { name: 'View case' }).click();
> For this I get the error: "Object is possibly 'null'.ts(2531)"

Indeed, but this is on one hand a TypeScript linting error and on the other hand not recommended anymore, since [FrameLocators](https://playwright.dev/docs/api/class-page#page-frame-locator) are superseding the `page.frame` API for most of the common scenarios. They will then get not only an easier way of selecting an element, but they will also get auto-retries out of the box for all the actions and assertions where they are used.

**Comment by ufst-ask** [comment]
Okay, but it was the code generated by codegen. Does this mean that codegen will use FrameLocators in a future release then?
And as long as codegen isn't 100% correct all the time, which I wouldn't expect it to be, I would still like to make the feature request.
Then if you don't agree with me that it's worthwhile with this find_in_page functionality then you can always decline :)

**Comment by mxschmitt** [comment]
Sounds like a bug! We should improve on that what I see in the code then. We already use in most of the cases FrameLocators, this specific case seems unexpected.

The find_in_page functionality seems outside of the scope for Playwright as per https://github.com/microsoft/playwright/issues/27650#issuecomment-1775987996. But we are more than happy to dig into bugs for codegen. The difference between our codegen and other codegen alternatives are that its actually performing the clicks via Playwright itself while you interact with the page. This ensures that the generated code works, except in rare cases e.g. hover.

CP (for me): https://github.com/microsoft/playwright/blob/cffb6ac269ab97ac63ac98d88244603e3349b793/packages/playwright-core/src/server/recorder/javascript.ts#L63

**Comment by ufst-ask** [comment]
> The find_in_page functionality seems outside of the scope for Playwright

Not the answer I hoped for but I think I understand why. And I appreciate the strength of codegen and your continous improvements on it :)
